include/seqan3/container/detail/ra_iterator.hpp

```diff
@@ -0,0 +1,217 @@
+
+#pragma once
+
+#include <iomanip>
+#include <limits>
+#include <type_traits>
+
+#include <seqan3/container/concepts.hpp>
+
+namespace seqan3::detail
+{
+
+template <typename container_type, bool is_const=false>
+    requires random_access_range_concept<container_type>
+struct ra_iterator
+{
+
+private:
+    //! Note: an iterator can only be initialized with container (_host) reference
+    container_type & _host{*(container_type*)0};
+    //! set to numeric_limits::max to indicate that iterator exceeds container size
+    typename container_type::size_type _pos{0};
+public:
+    typedef typename container_type::size_type size_type;
+    typedef typename container_type::difference_type difference_type;
+    typedef typename container_type::value_type value_type;
+    using reference = std::conditional_t<is_const, value_type const &, value_type &>;
+    typedef typename container_type::const_reference const_reference;
+    using pointer = std::conditional_t<is_const, value_type const *, value_type *>;
+
+    //! Explicitly not possible: ra_iterator() : _host(*(container_type*)0) {};
+    ra_iterator(container_type & _host, bool const _at_end = false) : _host{_host}, _pos{0}
+    {
+        if (_at_end)
+            _pos = std::numeric_limits<size_type>::max();
+    }
+
+    //! copy constructor
+    ra_iterator(ra_iterator const & _it) : _host{_it._host}, _pos{_it._pos} {}
+
+    //! copy via assignment
+    ra_iterator& operator=(ra_iterator const & rhs)
+    {
+        //assert(_host == rhs._host);
+        // TODO: remove above assert, because with assert: this is not possible:
+        // it = it + 2, but we allow it += 2; Hence, we would expect same behaviour.
+        _pos = rhs._pos;
+        return *this;
+    }
+
+    ~ra_iterator() = default;
+
+    //! return element currently pointed at
+    reference operator*() const
+    {
+        return _host[_pos];
+    }
+
+    //! two iterators are equal if their absolute positions are the same
+    bool operator==(ra_iterator const & rhs) const
+    {
+        return this->_pos == rhs._pos;
+    }
+
+    bool operator!=(ra_iterator const & rhs) const
+    {
+        return this->_pos != rhs._pos;
+    }
+
+    bool operator<(ra_iterator const & rhs) const
+    {
+        return this->_pos < rhs._pos;
+    }
+
+    bool operator>(ra_iterator const & rhs) const
+    {
+        return this->_pos > rhs._pos;
+    }
+
+    bool operator<=(ra_iterator const & rhs) const
+    {
+        return this->_pos <= rhs._pos;
+    }
+
+    bool operator>=(ra_iterator const & rhs) const
+    {
+        return this->_pos >= rhs._pos;
+    }
+
+    //! pre-increment, return updated iterator
+    ra_iterator& operator++()
+    {
+        ++_pos;
+        return (*this);
+    }
+
+    //! post-increment, return old iterator
+    ra_iterator operator++(int)
+    {
+        ra_iterator cpy{*this};
+        ++_pos;
+        if (_pos >= _host.size())
```

Comment:
wherever you call `host.size()` instead call `ranges::size(host)`, because this is the more generic version.
`.size()` is ok if we are dealing with containers only, but for general ranges it is only required that there is some free function. `ranges::size()` looks for a free function, then tries the member function and in the end subtracts the end and begin iterators so it is always a safe bet.